lib/datadog/tracing/trace_operation.rb

```diff
@@ -318,6 +323,22 @@ def subscribe_deactivate_trace(&block)
         :events,
         :root_span
 
+      def tags
```

Comment:
Alright, I tried that in [90614bb](https://github.com/DataDog/dd-trace-rb/pull/1959/commits/90614bb9cc54b0a2daebcea7840d852b572c36ce) but I experience cognitive dissonance about it:

- it exposes `set_error` as well, which does add error tags to the trace, but the error count stays `0` on the span.
- suddenly there's a need to recompose `meta` + `metrics` to pass them to `TraceSegment.new`, which will then re-split them via `set_tags` (which calls `set_tag`)
- analytics special tags handling sounds nice and consistent, but it feels strange since IIUC the API and storage for that is supposed to be on the `TraceOperation` itself (`@sample_rate` notably doesn't get updated)

`Metadata::Analytics could not be included, but `Metadata::Tagging` still brings in `set_error`. I don't think errors make sense at the trace level, they belong to a specific span where the error occured. In addition, the meta/metrics split currently feels wrong at this level as well.

I feel more comfortable introducing `set_tag` right now - and maybe `set_metric`, but on the span `set_tag` handles both already, which will be catered for anyway when the final spans are generated, and since one is not supposed to publicly read from - by itself, doubly so if we make them private for now.

Reply:
Loic asked me to take a peek and yeah there's a bit of an interesting design tension here between reusing and complexity. `Datadog::Tracing::Metadata` is tantalizingly close to what we want, but not quite.

Since the actual duplication here is around ~15 lines of code (+ specs), vs just `include Datadog::Tracing::Metadata`, it seems to me that living with the duplication may be a reasonable trade-off, as otherwise there's quite a lot of added complexity coming in from the `Metadata` module, just to avoid 15 very straightforward lines of Ruby code.

Another approach, if we wanted to avoid the duplication, would be to extract a class, let's call it `TagKeeper`, which would expose the common tags APIs, and that then both `TraceOperation` and users of `Datadog::Tracing::Metadata` could make use of (by delegating most, but not all of the time).

Reply:
I don't think its adding complexity, I think reusing the module significantly reduces it. And given that tagging is a core behavior (not some extraneous detail) it makes more sense for it to be standardized.

Where I agree is that `TraceOperation` doesn't need everything in `Tagging` right now. I think `Tagging` should be split into smaller modules: `Tagging` (`set_tag`/`set_metric`) and `Errors` (`set_error`). Then `TraceOperation` can re-use the relevant `Tagging` module.

Reply:
Done